SNOW-1437885  Disable blob interleaving when in Iceberg mode

sfc-gh-alhuang opened the pull request:
* Added a method to `SnowflakeStreamingIngestClientInternal` to activate/deactivate streaming to Iceberg tables.
* A blob should only contain one chunk (parquet) under Iceberg mode. Disable blob interleaving under Iceberg mode.

sfc-gh-gdoci commented on src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java:
```diff
@@ -123,6 +123,10 @@ List<List<ChannelData<T>>> getData() {
 
+  // Indicates if it's flushing to Iceberg tables, a blob could only contain one chunk under Iceberg
+  // mode
+  private final boolean isIcebergMode;
```
Can this have a more general name, e.g. `isNonInterleavedMode`?

sfc-gh-tzhang replied:
Yeah, we can make this more general for the internal names since we disable interleaved mode due to other reasons in the future, not only Iceberg

sfc-gh-alhuang replied:
Removed the non-interleave parameter for now as we could control this via `MAX_CHUNKS_IN_BLOB_AND_REGISTRATION_REQUEST`

sfc-gh-hmadan commented on src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java:
```diff
@@ -437,7 +447,7 @@ && shouldStopProcessing(
           // Add processed channels to the current blob, stop if we need to create a new blob
           blobData.add(channelsDataPerTable.subList(0, idx));
-          if (idx != channelsDataPerTable.size()) {
+          if (idx != channelsDataPerTable.size() || isIcebergMode) {
```
I want to guard against the possibility of having isIcebergMode peppered across 10 different places in code, making it that much harder to maintain the client.

Ideally we should have one place in the whole client that does `if (isIcebergMode) { /* initialize booleans / ints / settings one way */ } else { /* initialize another way */ }` - it seems that the ParameterProvider is the best place to do this, or we introduce a separate ClientSettings class that holds all these settings that control different behaviors in different places in the client.

I also see there are three parameters on the parameter provider that need different values for iceberg - MAX_CHUNK_SIZE_IN_BYTES (set to 512 MB rn) and MAX_CHUNKS_IN_BLOB_AND_REGISTRATION_REQUEST_DEFAULT (set to 100 rn) and MAX_BLOB_SIZE_IN_BYTES (set to 1 GB rn). This is in addition to MAX_CLIENT_LAG needing an iceberg-specific value.

It looks like if parameterProvider.getMaxChunksInBlobAndRegistrationRequest() returns 1 then you'll get this same desired behavior ?

Note - in the parameterProvider you'll need to validate that customers can't override MAX_CHUNKS_IN_BLOB_AND_REGISTRATION_REQUEST and it can only ever be 1.

sfc-gh-alhuang replied:
Deleted the `isIcebergMode` param and use `MAX_CHUNKS_IN_BLOB_AND_REGISTRATION_REQUEST` for now.

sfc-gh-hmadan commented on src/test/java/net/snowflake/ingest/streaming/internal/FlushServiceTest.java:
```diff
@@ -422,7 +427,7 @@ public void testGetFilePath() {
   @Test
   public void testFlush() throws Exception {
-    TestContext<?> testContext = testContextFactory.create();
+    TestContext<?> testContext = testContextFactory.create(false);
```
as suggested in another thread please using JUnit parameters / somehow parameterize this isIcebergMode so that all tests run with isIcebergMode=on/off, instead of only testing the non-iceberg usecases.

feel free to close this thread out so we just have one thread on this!

sfc-gh-hmadan reviewed (changes requested):
Left a couple comments, PTAL.

sfc-gh-tzhang commented on src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestClientInternal.java:
```diff
@@ -258,16 +265,17 @@ public SnowflakeStreamingIngestClientInternal(
    */
   SnowflakeStreamingIngestClientInternal(String name) {
-    this(name, null, null, null, true, null, new HashMap<>());
+    this(name, null, null, null, false, true, null, new HashMap<>());
```
agree, you can search for `Parameterized` and see how to use that

sfc-gh-alhuang replied:
Added parameters for unit tests.

sfc-gh-tzhang reviewed (commented):
I think there're a few things missing in this PR:
- Block creation of regular channels on Iceberg client and creation of iceberg channels on regular client
- No need to call the client/configure endpoint for Iceberg client

Do you plan to do that in a separate PR?

sfc-gh-hmadan commented on src/main/java/net/snowflake/ingest/streaming/SnowflakeStreamingIngestClientFactory.java:
```diff
@@ -28,6 +28,9 @@ public static class Builder {
     private Map<String, Object> parameterOverrides;
 
+    // Indicates whether it's streaming to Iceberg tables
+    private boolean isIcebergMode;
```
I am adding this to the server side validation in another PR already; do we typically do open-channel validations on both sides?

sfc-gh-alhuang commented on src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java:
```diff
@@ -437,7 +447,7 @@ && shouldStopProcessing(
           // Add processed channels to the current blob, stop if we need to create a new blob
           blobData.add(channelsDataPerTable.subList(0, idx));
-          if (idx != channelsDataPerTable.size()) {
+          if (idx != channelsDataPerTable.size() || isIcebergMode) {
```
Removed the non-interleave parameter for now as we could control this via MAX_CHUNKS_IN_BLOB_AND_REGISTRATION_REQUEST

sfc-gh-alhuang commented:
Thanks for the comments and tips! Added `icebergMode` on/off params for some unit tests. Some tests included blob registration / authentication only test with non regular mode for now. @sfc-gh-tzhang I plan to delete/modify the client configure call in later PRs as it's only called by `StreamingIngestStage` and we might not using this class in the iceberg mode later. The Iceberg client/server check is done in the server side ([ref](https://github.com/snowflakedb/snowflake-ingest-java/pull/763#discussion_r1617601137)), should we also added check in the client?

sfc-gh-hmadan commented on src/main/java/net/snowflake/ingest/utils/ParameterProvider.java:
```diff
@@ -188,6 +200,13 @@ private void setParameterMap(Map<String, Object> parameterOverrides, Properties
+    // Required parameter override for Iceberg mode
+    if (this.isIcebergMode) {
+      this.parameterMap.put(
+          MAX_CHUNKS_IN_BLOB_AND_REGISTRATION_REQUEST,
```
why not call this.updateValue here ? That's the only method that's updating parameterMap before this block was added, lets keep it that way..?

sfc-gh-alhuang replied:
The `updateValue` considers props and paramOverrides before default value. While under iceberg mode we should not allow user to set the chunk number > 1, that's why I use the `put` method directly. I think we can use `updateValue(key, value, null, null)` to achieve the same result, wdyt?

sfc-gh-hmadan reviewed (approved):
LGTM, left one small comment that you can take in the next PR too as you have other changes waiting for this to go in.

sfc-gh-tzhang commented on src/main/java/net/snowflake/ingest/utils/ParameterProvider.java:
```diff
@@ -64,6 +64,12 @@ public class ParameterProvider {
+  public static final int MAX_CHUNKS_IN_BLOB_AND_REGISTRATION_REQUEST_ICEBERG_MODE_DEFAULT = 1;
+
+  // If the provided parameters need to be verified and modified to meet Iceberg mode
+  private final boolean isIcebergMode;
```
what happens if customer create a Iceberg client but override this value to false?

sfc-gh-alhuang replied:
I don't think client can inject `ParameterProvider` directly, instantiate is called by `ClientInternal` at [here](https://github.com/snowflakedb/snowflake-ingest-java/blob/6c9bc9b3e3b9bd687898aeb59cb0d30940b4c8c9/src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestClientInternal.java#L172).

sfc-gh-tzhang commented on src/main/java/net/snowflake/ingest/utils/ParameterProvider.java:
```diff
@@ -64,6 +64,12 @@ public class ParameterProvider {
       Constants.BdecParquetCompression.GZIP;
 
+  // Iceberg mode default parameters
+  public static final int MAX_CHUNKS_IN_BLOB_AND_REGISTRATION_REQUEST_ICEBERG_MODE_DEFAULT = 1;
```
This sounds like we're allowing to change it to something other than 1, but it has to be one for Iceberg, right? We should put this into Constants.java, everything in this file means that we're allow to override it.

sfc-gh-alhuang replied:
Yes. Moved to `Constants.java`.

sfc-gh-tzhang commented on src/test/java/net/snowflake/ingest/streaming/internal/FlushServiceTest.java:
```diff
@@ -48,11 +48,22 @@
+@RunWith(Parameterized.class)
 public class FlushServiceTest {
+
+  @Parameterized.Parameters(name = "isIcebergMode: {0}")
```
I thought what we should be doing is to test the isIcebergMode with {true, false}

sfc-gh-alhuang replied:
This line is the format of the test name, where the first parameter is either `false` or `true` from [here](https://github.com/snowflakedb/snowflake-ingest-java/blob/6c9bc9b3e3b9bd687898aeb59cb0d30940b4c8c9/src/test/java/net/snowflake/ingest/streaming/internal/FlushServiceTest.java#L62).

sfc-gh-alhuang commented on src/main/java/net/snowflake/ingest/streaming/SnowflakeStreamingIngestClientFactory.java:
```diff
@@ -45,6 +49,11 @@ public Builder setParameterOverrides(Map<String, Object> parameterOverrides) {
       return this;
     }
 
+    public Builder setIcebergMode(boolean isIcebergMode) {
```
Renamed.

sfc-gh-alhuang commented:
Per offline discussion with @sfc-gh-tzhang , moved `MAX_CHUNKS_IN_BLOB_AND_REGISTRATION_REQUEST_ICEBERG_MODE_DEFAULT` back to ParameterProvider.java.

sfc-gh-tzhang commented on src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestClientInternal.java:
```diff
@@ -258,16 +264,17 @@ public SnowflakeStreamingIngestClientInternal(
       SnowflakeURL accountURL,
       Properties prop,
       Map<String, Object> parameterOverrides,
+      boolean isIcebergMode,
```
Let's make sure this is being added to the comment as well

sfc-gh-tzhang commented on src/main/java/net/snowflake/ingest/utils/ParameterProvider.java:
```diff
@@ -64,6 +66,9 @@ public class ParameterProvider {
   public static final Constants.BdecParquetCompression BDEC_PARQUET_COMPRESSION_ALGORITHM_DEFAULT =
       Constants.BdecParquetCompression.GZIP;
 
+  // Iceberg mode parameters
```
Let's extend this comment explain why we need a different default for iceberg

sfc-gh-tzhang commented on src/main/java/net/snowflake/ingest/utils/ParameterProvider.java:
```diff
@@ -411,4 +433,15 @@ public Constants.BdecParquetCompression getBdecParquetCompressionAlgorithm() {
+
+  private void icebergModeValidation(String key, Object expected) {
+    Object val = this.parameterMap.get(key);
+    if (!val.equals(expected)) {
```
We should throw an exception instead of just a warning

sfc-gh-tzhang reviewed (approved):
Left some comments, PTAL, otherwise LGTM! One request I have is could we merge all the Iceberg related changes to a feature branch instead of `main`? We can move everything to `main` once we ready, otherwise the new release will contain all the WIP changes and customer might do something with it.
